src/main/scala/firrtl/ir/IR.scala

```diff
@@ -211,8 +211,28 @@ abstract class Width extends FirrtlNode {
   }
 }
 /** Positive Integer Bit Width of a [[GroundType]] */
-case class IntWidth(width: BigInt) extends Width {
+object IntWidth {
+  private val maxCached = 1024
+  private val cache = new Array[IntWidth](maxCached)
+  def apply(width: BigInt): IntWidth = {
+    if (1 <= width && width <= maxCached) {
+      val i = width.toInt - 1
+      var w = cache(i)
+      if (w eq null) {
+        w = new IntWidth(width) ; cache(i) = w
+      }
+      w
+    } else new IntWidth(width)
+  }
+  // For pattern matching
+  def unapply(w: IntWidth): Option[BigInt] = Some(w.width)
+}
+class IntWidth(val width: BigInt) extends Width {
```

Comment:
You generally need to override hashCode when you override equals. The default hashCode is the Java Object default which is basically hashing the memory location. This is improper for a class trying to establish value equality. The following (Java) article explains why (and, although not seemingly a problem here, other pitfalls): http://www.artima.com/lejava/articles/equality.html

Since other pull requests are indicating 0-width UInts are likely to be a better supported thing, it would probably make sense to just cache 0-width IntWidth as well. Also avoids needing to adjust the index. (Make sure to then change the check to < maxCached or +1 the array size.)

To complete the case class behavior you should also override toString and provide a copy method. I think the toString implementation usually uses parts of the Product extension Jack mentioned. I believe case classes also extend Serializable but am having trouble finding details on that.